image-rs/src/image.rs

```diff
@@ -336,10 +353,119 @@ impl ImageClient {
             .lock()
             .await
             .image_db
-            .insert(image_data.id.clone(), image_data);
+            .insert(image_data.id.clone(), image_data.clone());
 
         Ok(image_id)
     }
+
+    #[cfg(feature = "nydus")]
+    async fn do_pull_image_with_nydus<'a>(
+        &mut self,
+        client: &mut PullClient<'_>,
+        image_data: &mut ImageMeta,
+        image_manifest: &OciImageManifest,
+        decrypt_config: &Option<&str>,
+        bundle_dir: &Path,
+    ) -> Result<String> {
+        let diff_ids = image_data.image_config.rootfs().diff_ids();
+        let bootstrap_id = if !diff_ids.is_empty() {
+            diff_ids[diff_ids.len() - 1].to_string()
+        } else {
+            bail!("Failed to get bootstrap id, diff_ids is empty");
+        };
+
+        let layer_metas = client
+            .pull_bootstrap(
+                utils::get_nydus_bootstrap_desc(image_manifest).unwrap(),
+                bootstrap_id.to_string(),
+                decrypt_config,
+                self.meta_store.clone(),
+            )
+            .await?;
+        image_data.layer_metas = layer_metas;
+        let layer_db: HashMap<String, LayerMeta> = image_data
+            .layer_metas
+            .iter()
+            .map(|layer| (layer.compressed_digest.clone(), layer.clone()))
+            .collect();
+
+        self.meta_store.lock().await.layer_db.extend(layer_db);
+
+        if image_data.layer_metas.is_empty() {
+            bail!("Failed to pull the bootstrap");
+        }
+
+        let reference = Reference::try_from(image_data.reference.clone())?;
+        let nydus_config = self
+            .config
+            .get_nydus_config()
+            .expect("Nydus configuration not found");
+        let work_dir = self.config.work_dir.clone();
+
+        let image_id = service::start_nydus_service(
```

Comment:
Do we need start nydus service for each `pull_image` request? or create one nydus service daemon for each `ImageClient` instance?

Reply:
One nydus daemon for one image reference. If we want to use image-rs to download a new image, then we need to reconfigure the nydus daemon and start a new one. If we download the same image, there is no need to start a new daemon. One nydus daemon for all images is unsupported.